### ISSUE TITLE
NoMethodError: private method `open' called for #<ActiveStorage::Downloader:0x00007f024fa16078>

### ISSUE THREAD

**Issue by ttibensky** [rails/rails#39798]
### Steps to reproduce

I'm executing `RAILS_ENV=test bin/rails db:setup` which should populate my test database with some fake data. After I upgraded from rails 5.2 to 6.0.3.2, this rake no longer works. It looks like it is trying to move the provided image to a temporary location which fails because a private method has been called. However, the `open()` method is not being called from our code directly.

### seeds.rb
```ruby
category = Category.create(name: 'My category')

4.times do |index|
  organization = Organization.new({
                                    name: "Organization #{index+1}",
                                    address: 'here',
                                    city: 'Chicago',
                                    state: 'IL',
                                    url: 'https://organization.com',
                                    description: 'This is description',
                                    ein: '12345'})
  default_community = Community.new({ name: "Default community for Organization #{index+1}", description: 'desc', organization: organization })
  organization.default_community = default_community

  organization.profile_photo.attach(io: File.open(Rails.root.join('spec', 'factories', 'files', 'organization_photo.jpg')), filename: 'organization_photo.jpg', content_type: 'image/jpeg')
  organization.white_profile_photo.attach(io: File.open(Rails.root.join('spec', 'factories', 'files', 'organization_photo.jpg')), filename: 'organization_photo.jpg', content_type: 'image/jpeg')
  organization.cover_photo.attach(io: File.open(Rails.root.join('spec', 'factories', 'files', 'organization_photo.jpg')), filename: 'organization_photo.jpg', content_type: 'image/jpeg')
  organization.categories << category

  organization.save! # this is -> /var/www/app_name/db/seeds.rb:23:in `block in <main>'
end
```

### Expected behavior
rake should finish normally

### Actual behavior
```bash
NoMethodError: private method `open' called for #<ActiveStorage::Downloader:0x00007f024fa16078>
/usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/service.rb:86:in `open'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/models/active_storage/blob.rb:219:in `open'
/usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/analyzer.rb:27:in `download_blob_to_tempfile'
/usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/analyzer/image_analyzer.rb:32:in `read_image'
/usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/analyzer/image_analyzer.rb:21:in `metadata'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/models/active_storage/blob/analyzable.rb:47:in `extract_metadata_via_analyzer'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/models/active_storage/blob/analyzable.rb:29:in `analyze'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/jobs/active_storage/analyze_job.rb:11:in `perform'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/execution.rb:40:in `block in perform_now'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:112:in `block in run_callbacks'
/usr/local/bundle/gems/i18n-1.8.3/lib/i18n.rb:313:in `with_locale'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/translation.rb:9:in `block (2 levels) in <module:Translation>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/core_ext/time/zones.rb:66:in `use_zone'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/timezones.rb:9:in `block (2 levels) in <module:Timezones>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:25:in `block (4 levels) in <module:Logging>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in `block in instrument'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in `instrument'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:24:in `block (3 levels) in <module:Logging>'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in `block in tag_logger'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in `block in tagged'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:28:in `tagged'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in `tagged'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in `tag_logger'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:21:in `block (2 levels) in <module:Logging>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:139:in `run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/execution.rb:39:in `perform_now'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/execution.rb:25:in `block in execute'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:112:in `block in run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/railtie.rb:43:in `block (4 levels) in <class:Railtie>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/execution_wrapper.rb:88:in `wrap'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/reloader.rb:72:in `block in wrap'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/execution_wrapper.rb:88:in `wrap'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/reloader.rb:71:in `wrap'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/railtie.rb:42:in `block (3 levels) in <class:Railtie>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:139:in `run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/execution.rb:23:in `execute'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/queue_adapters/inline_adapter.rb:15:in `enqueue'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/enqueuing.rb:59:in `block in enqueue'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:112:in `block in run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in `block in instrument'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/notifications.rb:180:in `instrument'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:35:in `block (2 levels) in <module:Logging>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:16:in `block (3 levels) in <module:Logging>'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in `block in tag_logger'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in `block in tagged'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:28:in `tagged'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/tagged_logging.rb:80:in `tagged'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:45:in `tag_logger'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/logging.rb:15:in `block (2 levels) in <module:Logging>'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:139:in `run_callbacks'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/enqueuing.rb:55:in `enqueue'
/usr/local/bundle/gems/activejob-6.0.3.2/lib/active_job/enqueuing.rb:22:in `perform_later'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/models/active_storage/blob/analyzable.rb:37:in `analyze_later'
/usr/local/bundle/gems/activestorage-6.0.3.2/app/models/active_storage/attachment.rb:37:in `analyze_blob_later'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:428:in `block in make_lambda'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:263:in `block in conditional'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:517:in `block in invoke_after'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:517:in `each'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:517:in `invoke_after'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:136:in `run_callbacks'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/callbacks.rb:825:in `_run_commit_callbacks'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/transactions.rb:339:in `committed!'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:127:in `commit_records'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:265:in `block in commit_transaction'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:255:in `commit_transaction'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:293:in `block in within_new_transaction'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:278:in `within_new_transaction'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `transaction'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/transactions.rb:212:in `transaction'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/transactions.rb:366:in `with_transaction_returning_status'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/transactions.rb:318:in `save!'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/suppressor.rb:48:in `save!'
/var/www/app_name/db/seeds.rb:23:in `block in <main>'
/var/www/app_name/db/seeds.rb:6:in `times'
/var/www/app_name/db/seeds.rb:6:in `<main>'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:55:in `load'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/engine.rb:557:in `block in load_seed'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/engine.rb:675:in `with_inline_jobs'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/engine.rb:557:in `load_seed'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/tasks/database_tasks.rb:440:in `load_seed'
/usr/local/bundle/gems/activerecord-6.0.3.2/lib/active_record/railties/databases.rake:331:in `block (2 levels) in <main>'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/commands/rake/rake_command.rb:23:in `block in perform'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/commands/rake/rake_command.rb:20:in `perform'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/command.rb:48:in `invoke'
/usr/local/bundle/gems/railties-6.0.3.2/lib/rails/commands.rb:18:in `<main>'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
/usr/local/bundle/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:324:in `block in require'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:291:in `load_dependency'
/usr/local/bundle/gems/activesupport-6.0.3.2/lib/active_support/dependencies.rb:324:in `require'
/var/www/app_name/bin/rails:9:in `<top (required)>'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/client/rails.rb:28:in `load'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/client/rails.rb:28:in `call'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/client/command.rb:7:in `call'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/client.rb:30:in `run'
/usr/local/bundle/gems/spring-2.1.0/bin/spring:49:in `<top (required)>'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/binstub.rb:11:in `load'
/usr/local/bundle/gems/spring-2.1.0/lib/spring/binstub.rb:11:in `<top (required)>'
/var/www/app_name/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
Tasks: TOP => db:setup => db:seed
(See full trace by running task with --trace)
```

### System configuration
**Rails version**: 6.0.3.2

**Ruby version**: 2.7.1 (docker image ruby:2.7.1)

**Comment by victorperez** [comment]
Can you create a demo application that reproduces the issue?

I'm trying to reproduce the error using a rails `6.0.3.2` using a similar script but no error is raised.

**Comment by ttibensky** [comment]
@victorperez It will be a bit difficult to create a demo app, but I will try if we can't figure out anything else.

So far, I discovered a bit more of the weird behavior:
```ruby
    # /usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/service.rb
    def open(*args, **options, &block)
      puts ActiveStorage::Downloader.new(self).inspect
      exit
      ActiveStorage::Downloader.new(self).open(*args, **options, &block)
    end
```
```bash
$ RAILS_ENV=test bin/rails db:setup
#<ActiveStorage::Downloader:0x00007fd936d1e4c8 @blob=#<ActiveStorage::Service::DiskService:0x00007fd9372be3d8 @root="/var/www/app_name/tmp/storage">, @tempdir=nil>
```
this looks all good, right? But let's see what happens when I add another method to Downloader class and try to call it statically:
```ruby
    # /usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/service.rb
    def open(*args, **options, &block)
      ActiveStorage::Downloader::lala
      exit
      ActiveStorage::Downloader.new(self).open(*args, **options, &block)
    end
```
```ruby
module ActiveStorage
  class Downloader #:nodoc:
    attr_reader :service

    def initialize(service)
      @service = service
    end

    def self.lala
      puts 'lala'
    end
...
```
```bash
$ RAILS_ENV=test bin/rails db:setup
NoMethodError: undefined method `lala' for ActiveStorage::Downloader:Class
/usr/local/bundle/gems/activestorage-6.0.3.2/lib/active_storage/service.rb:86:in `open'
```
the same happens if I call the method on an instance

**Comment by ttibensky** [comment]
I tried to disable the analyzers by putting `config.active_storage.analyzers = [ActiveStorage::Analyzer::NullAnalyzer]` to `config/application.rb` based on [official docs](https://railsguides.kr/configuring.html#configuring-active-storage) but it looks like this setting is ignored.

**Comment by ttibensky** [comment]
@victorperez I tried to replicate the issue in the demo app without any luck, it always works in the demo. Took us a few days to notice that the whole **Download class is overridden** in our project files... After I upgraded from rails 5.2 to rails 6 it stopped working because some modules/classes/methods in the libraries have changed and our overridden class stopped working. The fix was just to delete the overridden class since the functionality is not needed anymore anyway. Thank you for your time and quick response :) I'm going to close this.
